lib/python/qmk/cli/pid.py

```diff
@@ -0,0 +1,196 @@
+            if id_pid is None:
+                id_pid = re.match(PID_regex, line)
+            if id_vid is not None and id_pid is not None:
+                break
+    return id_vid, id_pid
+
+
+def calculate_pid(config_path, data, max_tries=10):
+    """Calculate a PID given the path to a config.h
+
+    This function generates a unique PID for any config.h supplied.
+    The PID is determined based on the SHA1 hexdigest of the path to the config file
+    The first 4 characters (2 bytes) are used as the PID, unless those collide with
+    an already assigned PID. If a collision happens the process will be repeated with offset + 1
+
+    Args:
+        config_path (str): The path to the config file to process.
+        data (dict):
+
+            * PID (str): The parent folder of the config associated with the PID key
+
+        max_tries (int): The maximum amount of attempts to get around collisions
+
+    Returns:
+        new_pid (str): The 4 characters of the 2 byte pid
+    """
+    m = hashlib.sha1()
+    m.update(config_path.encode('utf-8'))
+    digest = m.hexdigest()
+    offset = 0
+    while offset < max_tries:
+        id_pid = digest[offset:offset + 4].upper()
+
+        if id_pid not in data['pids']:
+            return id_pid
+        else:
+            offset += 1
+
+    raise RecursionError(offset)
+
+
+def atomic_dump(data, json_file):
+    """Dumps a dict as a json file atomically
+
+    This function should avoid data loss caused by an error happening during
+    file writing.
+    It accomplishes this by writing the new json file to a temporary file,
+    then using os.replace() to overwrite the old file with the new file.
+    os.replace() is atomic as long as the two files are on the same filesystem.
+    In case of an error before os.replace() the temp file isn't removed.
+
+    Args:
+        data (dict):
+
+            * PID (str): The parrent folder of the config associated with the PID key
+
+        json_file (str): The file to dump to
+    """
+    with tempfile.NamedTemporaryFile('w', dir=str(Path(json_file).parents[0]), delete=False) as tf:
+        json.dump(data, tf)
+    os.replace(tf.name, json_file)
+
+
+def init(pids_json_path):
+    """On the first run of this script, write a skeleton json file"""
+    Path(pids_json_path).parents[0].mkdir(parents=True, exist_ok=True)
+    with open(pids_json_path, 'w') as jfile:
+        json.dump({"pids": {}}, jfile)
+
+
+def replace_pid(config_h, id_pid, pid_match):
+    """Replace the PID in a config.h file
+
+    Args:
+        config_h (str): The config to replace the PID in. The file has to have a "#define PRODUCT_ID 0x0000" the value doesn't matter, but has to be 2 bytes
+        id_pid (str): The PID to write in to the file
+        pid_match (re.match): The full re match for the PID string
+    """
+
+    for line in fileinput.input(config_h, inplace=True):
+        print(line.replace(pid_match.group(0), "{}{}".format(pid_match.group(0)[:-4], id_pid)), end='')
```

Comment:
One weakness of this script right now, is that it relies on configs to have `#define PRODUCT_ID 0x0000` in the config. The specific hash value isn't important, but it is important that it's 4 characters, as this line replaces the last 4 characters in that string with the generated PID. If this isn't present the regex will fail and the script will error out ("nicely")

It's done this way to ensure that the PID is indented the same as it was in the original file by the author. Just wanted to comment on it to make sure it wasn't missed.

Reply:
More info here:

If you change the regex to my suggestion, you'll have 2 capture groups. `group(1)` will contain everything except `0x0000` and `group(2)` will contain just that.
This way, you can just replace the line with `"%s%s" % (pid_match(group2), id_pid)` and the original whitespacing will be preserved.

Reply:
Done, thanks! :)